mongo_test.go

```diff
@@ -202,248 +246,352 @@ func TestDelete(t *testing.T) {
-			}
+	t.Run("items should "+tt.ItemsShould, func(t *testing.T) {
+		if got := len(l.Items); got != tt.ExpectItemsLength {
+			t.Errorf("Unexpected result for len(l.Items):\nexpect: %d\ngot: %d", tt.ExpectItemsLength, got)
 		}
-	}
-
-	q, err = query.New("", `{name:{$regex:"^re[s]{1}t-.+yer.+exp$"}}`, "", query.Page(1, 1, 0))
-	if assert.NoError(t, err) {
-		l, err = h.Find(ctx, q)
-		if assert.NoError(t, err) {
-			assert.Equal(t, -1, l.Total)
-			if assert.Len(t, l.Items, 1) {
-				item := l.Items[0]
-				assert.Equal(t, "5", item.ID)
-				assert.Equal(t, map[string]interface{}{"id": "5", "name": "rest-layer-regexp"}, item.Payload)
-			}
+		if tt.CheckItems && !reflect.DeepEqual(l.Items, tt.ExpectItems) {
+			t.Errorf("Unexpected result Items slice:\nexpect: %v\ngot: %v", tt.ExpectItems, l.Items)
 		}
-	}
+	})
+}
 
-	q, err = query.New("", `{id:"10"}`, "", query.Page(1, 1, 0))
-	if assert.NoError(t, err) {
-		l, err = h.Find(ctx, q)
-		if assert.NoError(t, err) {
-			assert.Equal(t, 0, l.Total)
-			assert.Len(t, l.Items, 0)
-		}
+func TestFind(t *testing.T) {
+	s, cleanup := setupDBTest(t)
+	defer cleanup()
+	h := mongo.NewHandler(s, "", "test")
+	h2 := mongo.NewHandler(s, "", "test2")
+	allItems := []*resource.Item{
+		{ID: "1", Payload: map[string]interface{}{"id": "1", "name": "a", "age": 1}},
+		{ID: "2", Payload: map[string]interface{}{"id": "2", "name": "b", "age": 2}},
+		{ID: "3", Payload: map[string]interface{}{"id": "3", "name": "c", "age": 3}},
+		{ID: "4", Payload: map[string]interface{}{"id": "4", "name": "d", "age": 4}},
+		{ID: "5", Payload: map[string]interface{}{"id": "5", "name": "rest-layer-regexp"}},
 	}
+	if err := h.Insert(context.Background(), allItems); err != nil {
+		t.Fatalf("Unexpected error: %s", err)
+	}
+	if err := h2.Insert(context.Background(), allItems); err != nil {
+		t.Fatalf("Unexpected error: %s", err)
+	}
+
+	t.Run("when using an empty query", testFindOK{
+		h:     h,
+		Query: &query.Query{},
+
+		TotalShould: "count all items",
+		ExpectTotal: len(allItems),
+
+		ItemsShould:       "include all items",
+		ExpectItemsLength: len(allItems),
+		// Do not check result's content as its order is unpredictable.
+		CheckItems: false,
+	}.Run)
+
+	t.Run("when setting limit to 0", testFindOK{
+		h:     h,
+		Query: &query.Query{Window: &query.Window{Limit: 0}},
+
+		TotalShould: "count all items",
+		ExpectTotal: len(allItems),
+
+		ItemsShould:       "be an empty list",
+		ExpectItemsLength: 0,
+		ExpectItems:       []*resource.Item{},
+	}.Run)
+	t.Run("when setting limit -1 and offset to 2", testFindOK{
+		h:     h,
+		Query: &query.Query{Window: &query.Window{Limit: -1, Offset: 2}},
+
+		TotalShould: "count all items",
+		ExpectTotal: len(allItems),
+
+		ItemsShould:       "include all items except the first two",
+		ExpectItemsLength: len(allItems) - 2,
+		// Do not check result's content as its order is unpredictable.
+		CheckItems: false,
+	}.Run)
```

Comment:
@rs, are you OK with this style of tests?

This is a hybrid table tests and tests inspired by BDD-frameworks, so it looks a bit weird.

If dropping the table-aspect, I would write it as:

```go
 t.Run("when using an empty query", func(t *testing.T) {
    // do the query...
    t.Run("it should not error", func(t *testing.T) {
        if err != nil {
             t.Errorf("Unexpected error: %s", err)
        }
    })
    t.Run("total should count all items", func(t *testing.T) {
        // check...
    }
   t.Run("items should include all items", func(t *testing.T) {
        // check...
    }
})
```

However, in this case, at least for the [positive tests](http://www.aptest.com/glossary.html#positivetesting), it was a lot of repetition, and I found it easier to rewrite (and in the future maintian) if compromising by using a struct for the sub-tests. If there where any [negative tests](http://www.aptest.com/glossary.html#negativetesting), I would probably have written them without a helper-struct as shown above, or used a different helper struct if there was many of them, to avoid the if-else logic in the test runner.

It's inspired by BDD-style frameworks, where it's common to nest tests alongside textual descriptions of a behavior, and the expected consequences. E.g:

- "when doing X" // a condition
  - "Y should be A" // a check
  - "Z should be B" // another check

- "when doing X" // a condition
  - "when doing Y" // a sub-condition
    - "Z should be A" // a check

It may seam odd at first, but it really helps in documenting how the programmer things the code should work, and does a lot to test readability and maintainability as the purpose of the tests is better documented.

PS! BDD-friendly frameworks, like [Ginko](https://github.com/onsi/ginkgo) or [Spec](https://github.com/sclevine/spec), often wraps this with a good portion of magic so that the outer-code is rerun for every single check. In practice, I have found that closures can be used explicitly when such behavior is required. In this particular case (Find), it isn't required.

Reply:
What is the advantage of calling `Run` yourself as opposed to a real table test with the string of the run in the test struct and the run function in the test body for loop instead of a func of the test struct?

Reply:
Rewrote them to not have a common Run function, which make it closer to BDD, and further away from table tests. This way the checks can be chosen (and documented with an intent) differently for each test-case, while still allowing for short test-code and code reuse.

The line numbers for failures are now also finally reported correctly when `t.Helper()` is used for the check and test functions, which they oddly where not with the predefined Run function. In VS Code, this means you can now command-click on the failure to jump directly to the failing sub-test, making issues faster to correct.

Example test case:
```go

 	t.Run("when setting limit -1 and offset matching the length of all items +1", func(t *testing.T) {
 		l := doPositiveFindTest(t, h, &query.Query{
 			Window: &query.Window{Limit: -1, Offset: len(allItems)},
 		})
 
 		// Not able to get total count in one query.
 		t.Run("ItemsList.Total should not be deduced", totalCheckFunc(-1, l))
 		// Check that we get an empty Item list and not nil.
 		expectItems := []*resource.Item{}
 		t.Run("ItemList.Items should be an empty list", itemsCheckFunc(expectItems, l))
 	})
```

The changes are done as a separate commit for easier review. Will squash the commits before merge.

Reply:
Failed test in VS Code:

![image](https://user-images.githubusercontent.com/75205/50637445-5a6bd380-0f5a-11e9-98ea-a6859c168ff7.png)

Command click on the file / line number brings me to the place where the failing check function is called:

![image](https://user-images.githubusercontent.com/75205/50637479-6eafd080-0f5a-11e9-9d59-9f55db988109.png)